onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,709 @@
+      VLOGS(logger, 1) << "PropagateFP16CastsFromOutputsToInputs: Inserted Cast node to " << ConcatNames(node_args);
+      modified = true;
+    }
+  }
+  return modified;
+}
+
+// Expand FP16 compute regions on the graph by example float16 compute nodes,
+// propagating float32 Cast operation down the graph and propagating float16
+// Cast operations up the graph. The following functions are performed
+// 1. Fuse subgraphs
+// 2. Propagate fp32 casts forwards
+// 3. Propagate fp16 casts back
+Status PropagateCastOps::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  // First apply the transformation to the subgraphs.
+  {
+    GraphViewer graph_viewer(graph);
+    const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+    for (auto node_index : node_topology_list) {
+      auto* node_ptr = graph.GetNode(node_index);
+      if (nullptr == node_ptr)
+        continue;  // node was removed
+
+      auto& node = *node_ptr;
+
+      ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+    }
+  }
+  std::vector<std::string> removed_node_names;
+  int pass = 0;
+  bool local_modified = false;
+  do {
+    VLOGS(logger, 1) << "Propagate Cast Operations Pass " << pass << ":";
+    std::deque<onnxruntime::NodeIndex> removed_nodes;
+
+    if (local_modified) {
+      graph.Resolve();
+      local_modified = false;
+    }
+
+    GraphViewer graph_viewer(graph);
+    const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+    // Remove unnecessary casts
+    for (auto node_index : node_topology_list) {
+      Node* node = graph.GetNode(node_index);
+      if (node == nullptr ||
+          std::find(removed_nodes.begin(), removed_nodes.end(), node->Index()) != removed_nodes.end() ||
+          node->OpType() != "Cast") {
+        continue;
+      }
+      local_modified |= RemoveUnnecessaryCasts(graph, node, removed_nodes, logger);
+    }
+
+    // Fuse subgraphs, sibling Cast nodes with same input
+    for (auto node_index : node_topology_list) {
+      Node* node = graph.GetNode(node_index);
+      if (node == nullptr ||
+          std::find(removed_nodes.begin(), removed_nodes.end(), node->Index()) != removed_nodes.end()) {
+        continue;
+      }
+      local_modified |= FuseSiblingCasts(graph, node, removed_nodes, logger);
+    }
+
+    // Propagate FP32 Casts forward
+    for (auto node_index : node_topology_list) {
+      Node* node = graph.GetNode(node_index);
+      if (node == nullptr ||
+          std::find(removed_nodes.begin(), removed_nodes.end(), node->Index()) != removed_nodes.end() ||
+          !IsCastTo(node, TensorProto::FLOAT)) {
+        continue;
+      }
+      local_modified |= PropagateForwards(graph, node, removed_nodes, level_, logger);
+    }
+
+    // Remove back to back Casts, with FLOAT->FLOAT16 followed by FLOAT16->FLOAT, but not the other way.
+    for (auto node_index : node_topology_list) {
```

Comment:
nit, update comment 
"FLOAT->FLOAT16 followed by FLOAT16->FLOAT" is not very clear to a reader